mdp.1
.\" This is the groff documentation source for MDP
.\"
.\" Preview with: groff -man -Tascii mdp.1
.\"           or: man -l mdp.1
.\"
.
.
.TH MDP 1 "2016-04-02" "User Commands"
.SH NAME
mdp \- A command-line based
markdown presentation tool
.SH SYNOPSIS
.B mdp
.RI [ OPTION ].\|.\|.\|
.RI [ FILE ]
.
.SH DESCRIPTION
.B mdp
is a command-line program that allows you to make elegant presentations from
.B markdown formatted
.IR FILE s.
.PP
It is as easy as write your presentation content in the text editor of your
preference and launch the presentation from the command-line.
.
.SH OPTIONS
.SS "Input Control"
.TP
.IR FILE
The input file from which the presentation is read. If no file is specified,
or if the file name is
.BR \- ","
the presentation is read from standard input.
.SS "Output Control"
.TP
.BR \-c ", " \-\^\-nocodebg
Don't change the background color of code blocks.
.TP
.BR \-e ", " \-\^\-expand
Enable character entity expansion (e.g. '&gt;' becomes '>').
.TP
.BR \-f ", " \-\^\-nofade
Disable color fading in 256 color mode.
.TP
.BR \-i ", " \-\^\-invert
Swap black and white color.
.TP
.IR "\fB\-j\fP N" ", " "\fB\-\^\-jump\fP N"
Jump to slide number \fIN\fP.
.TP
.BR \-s ", " \-\^\-noslidenum
Do not show slide number at the bottom.
.TP
.BR \-t ", " \-\^\-notrans
Disable transparency in transparent terminal.
.TP
.BR \-x ", " \-\^\-noslidemax
Show slide number, but not total number of slides.
.
.SS "Miscellaneous Options"
.TP
.BR \-d ", " \-\^\-debug
Enable debug messages on STDERR. Add multiple times to increases debug level.
.TP
.BR \-h ", " \-\^\-help
Display usage message and exit.
.TP
.BR \-v ", " \-\^\-version
Display version and license information.
.
.SH ENVIRONMENT VARIABLES
.SS "Output Control"
.TP
.BR MDP_LIST_HEAD[1-3],\ MDP_LIST_OPEN[1-3]
Controls the list characters of unordered lists.

The default is equivalent to:
.br
MDP_LIST_OPEN1=' |  '
.br
MDP_LIST_OPEN2=' |  '
.br
MDP_LIST_OPEN3=' |  '
.br
MDP_LIST_HEAD1=' +- '
.br
MDP_LIST_HEAD2=' +- '
.br
MDP_LIST_HEAD3=' +- '
.
.SH MARKDOWN FORMATTING
For a complete list of supported markups, refer the sample presentation
(sample.md) provided alongside
.BR mdp ,\|
or online available at
.IR https://github.com/visit1985/mdp .
.SS "Slides"
The input
.IR FILE
is split into multiple slides by horizontal rules. Each consisting of at least
3
.B \-\-\-
or
.B ***
characters on a single line.
.B This line must be prefixed by an completely empty line.
It can also contain spaces but no other characters.
.PP
If any slide is too large to fit into your current screen, an error message
will be displayed at the moment the presentation is launched.
.
.SS "Line-by-Line Mode"
.SS "Block-by-Block Mode"
A single
.BR "<br>" ", " "<BR>" " or " "^"
on an otherwise empty line signals
.B mdp
to stop output of the current slide (stop point) and wait for a key-press by
the user.
.PP
This enables the user to display bullet points or list items one by one
(line by line) or block by block.
.
.SS "Headers"
.B mdp
supports header lines in the format of
.BR @ "[DESCRIPTION] " [VALUE]
The first two header lines are displayed as title and author in top and
bottom bar.
.PP
Headers are only recognized at the top of the input
.IR FILE .
.
.SS "Line spanning markup"
Supported are headlines, code blocks, quotes and unordered lists.
.
.SS "In-line markup"
As well as bold text, underlined text and in-line code.
.
.SH COLOR SUPPORT
Most terminals are able to display 256 colors these days. But some of them
enable only 16 colors by default. To enjoy
.BR mdp "'s"
full capabilities, these terminals need to be signaled to enable 256 color
mode. This is usually done by setting the TERM environment variable.
.PP
.BR "export TERM=xterm-256color"
.
.SH KEYBOARD CONTROLS
.TP
.BR "h, j, k, l, Arrow keys, Space, Enter, Backspace, Page Up, Page Down"
Display next/previous slide or continue after a stop point.
.TP
.BR "g, Home"
Jump to first slide.
.TP
.BR "G, End"
Jump to last slide.
.TP
.BR "1..N"
Jump to
.BR N "th"
slide.
.TP
.BR "r"
Reload the input
.IR FILE .\|
This key is disabled if input was read from standard input.
.TP
.BR "q"
Exit
.BR mdp "."
.
.SH CUSTOMIZATION
.B mdp
can be configured by modifying config.h and recompiling.
.SH AUTHOR
Written by Michael Goehler and others, see
.IR https://github.com/visit1985/mdp/blob/master/AUTHORS "."
.SH COPYRIGHT
Copyright (C) 2018 Michael Goehler
.PP
This is free software; see the source for copying conditions. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


mdp.c
/*
 * mdp -- A command-line based markdown presentation tool.
 * Copyright (C) 2014 Michael Goehler
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "include/mdp.h"

void usage() {
    fprintf(stderr, "%s", "Usage: mdp [OPTION]... [FILE]\n");
    fprintf(stderr, "%s", "A command-line based markdown presentation tool.\n\n");
    fprintf(stderr, "%s", "  -d, --debug     enable debug messages on STDERR\n");
    fprintf(stderr, "%s", "                  add it multiple times to increases debug level\n");
    fprintf(stderr, "%s", "  -f, --nofade    disable color fading in 256 color mode\n");
    fprintf(stderr, "%s", "  -h, --help      display this help and exit\n");
    fprintf(stderr, "%s", "  -i, --invert    swap black and white color\n");
    fprintf(stderr, "%s", "  -t, --notrans   disable transparency in transparent terminal\n");
    fprintf(stderr, "%s", "\nWith no FILE, or when FILE is -, read standard input.\n\n");
    exit(EXIT_FAILURE);
}

void version() {
    printf("mdp %d.%d.%d\n", MDP_VER_MAJOR, MDP_VER_MINOR, MDP_VER_REVISION);
    printf("Copyright (C) 2014 Michael Goehler\n");
    printf("License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n");
    printf("This is free software: you are free to change and redistribute it.\n");
    printf("There is NO WARRANTY, to the extent permitted by law.\n");
    printf("\nWritten by Michael Goehler and others, see <https://github.com/visit1985/mdp/blob/master/AUTHORS>.\n");
    exit(EXIT_SUCCESS);
}

int main(int argc, char *argv[]) {
    int notrans = 0;
    int nofade = 0;
    int invert = 0;

    // define command-line options
    struct option longopts[] = {
        { "debug",   no_argument, 0, 'd' },
        { "nofade",  no_argument, 0, 'f' },
        { "help",    no_argument, 0, 'h' },
        { "invert",  no_argument, 0, 'i' },
        { "notrans", no_argument, 0, 't' },
        { "version", no_argument, 0, 'v' },
        { 0, 0, 0, 0 }
    };

    // parse command-line options
    int opt, debug = 0;
    while ((opt = getopt_long(argc, argv, ":dfhitv", longopts, NULL)) != -1) {
        switch(opt) {
            case 'd': debug += 1; break;
            case 'f': nofade = 1; break;
            case 'h': usage(); break;
            case 'i': invert = 1; break;
            case 't': notrans = 1; break;
            case 'v': version(); break;
            case ':': fprintf(stderr, "%s: '%c' requires an argument\n", argv[0], optopt); usage(); break;
            case '?':
            default : fprintf(stderr, "%s: option '%c' is invalid\n", argv[0], optopt); usage(); break;
        }
    }

    // open file or set input to STDIN
    char *file;
    FILE *input;
    if (optind < argc) {
        do {
            file = argv[optind];
        } while(++optind < argc);

        if(!strcmp(file, "-")) {
            input = stdin;
        } else {
            input = fopen(file,"r");
            if(!input) {
                fprintf(stderr, "%s: %s: %s\n", argv[0], file, strerror(errno));
                exit(EXIT_FAILURE);
            }
        }
    } else {
        input = stdin;
    }

    // load deck object from input
    deck_t *deck;
    deck = markdown_load(input);

    // close file
    fclose(input);

    // replace stdin with current tty if input was a pipe
    if(input == stdin) {
        input = freopen("/dev/tty", "rw", stdin);
        if(!input) {
            fprintf(stderr, "%s: %s: %s\n", argv[0], "/dev/tty", strerror(errno));
            exit(EXIT_FAILURE);
        }
    }

    if(debug > 0) {
        markdown_debug(deck, debug);
    }

    ncurses_display(deck, notrans, nofade, invert);

    return(EXIT_SUCCESS);
}

